opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,370 @@
+/// The default max capacity for the stack portion of `GrowableArray`.
+const DEFAULT_MAX_INLINE_CAPACITY: usize = 10;
+/// The default initial capacity for the vector portion of `GrowableArray`.
+const DEFAULT_INITIAL_OVERFLOW_CAPACITY: usize = 5;
+
+#[derive(Debug, Clone, PartialEq)]
+/// A hybrid vector that starts with a fixed-size array and grows dynamically with a vector.
+///
+/// `GrowableArray` uses an internal fixed-size array (`initial`) for storing elements until it reaches
+/// `MAX_INLINE_CAPACITY`. When this capacity is exceeded, additional elements are stored in a heap-allocated
+/// vector (`additional`). This structure allows for efficient use of stack memory for small numbers of elements,
+/// while still supporting dynamic growth.
+///
+pub(crate) struct GrowableArray<
+    T: Default + Clone + PartialEq,
+    const MAX_INLINE_CAPACITY: usize = DEFAULT_MAX_INLINE_CAPACITY,
+    const INITIAL_OVERFLOW_CAPACITY: usize = DEFAULT_INITIAL_OVERFLOW_CAPACITY,
+> {
+    inline: [T; MAX_INLINE_CAPACITY],
+    overflow: Option<Vec<T>>,
+    count: usize,
+}
+
+impl<
+        T: Default + Clone + PartialEq,
+        const MAX_INLINE_CAPACITY: usize,
+        const INITIAL_OVERFLOW_CAPACITY: usize,
+    > Default for GrowableArray<T, MAX_INLINE_CAPACITY, INITIAL_OVERFLOW_CAPACITY>
+{
+    fn default() -> Self {
+        Self {
+            inline: [(); MAX_INLINE_CAPACITY].map(|_| T::default()),
+            overflow: None,
+            count: 0,
+        }
+    }
+}
+
+impl<
+        T: Default + Clone + PartialEq,
+        const MAX_INLINE_CAPACITY: usize,
+        const INITIAL_OVERFLOW_CAPACITY: usize,
+    > GrowableArray<T, MAX_INLINE_CAPACITY, INITIAL_OVERFLOW_CAPACITY>
+{
+    /// Creates a new `GrowableArray` with the default initial capacity.
+    #[allow(dead_code)]
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    /// Pushes a value into the `GrowableArray`.
+    ///
+    /// If the internal array (`inline`) has reached its capacity (`MAX_INLINE_CAPACITY`), the value is pushed
+    /// into the heap-allocated vector (`overflow`). Otherwise, it is stored in the array.    #[inline]
+    pub(crate) fn push(&mut self, value: T) {
+        if self.count < MAX_INLINE_CAPACITY {
+            self.inline[self.count] = value;
+            self.count += 1;
+        } else {
+            if self.overflow.is_none() {
+                // Initialize the vector with a specified capacity
+                self.overflow = Some(Vec::with_capacity(INITIAL_OVERFLOW_CAPACITY));
+            }
+            self.overflow.as_mut().unwrap().push(value);
```

Comment:
Can we use [get_or_insert_with](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with) here

Reply:
Yes, that's more idiomatic. Done.